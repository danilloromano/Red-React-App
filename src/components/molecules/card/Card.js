import React, { Component } from 'react';
import './Card.scss';
import { connect } from 'react-redux';
import cx from "classnames";
import PropTypes from "prop-types";

class Card extends Component {

    constructor(props) {
        super(props)

        this.state = {
            cardRotete: false,
            changeCard: false
        }

        this.formatCardNumber = this.formatCardNumber.bind(this)
    }

    formatCardNumber(number) {
        let numbers = number.toString()
        .replace(/(\d{0,4})(\d{0,4})(\d{0,4})(\d{0,4})/, "$1 $2 $3 $4");

        return {
            first: numbers.split(" ")[0],
            second: numbers.split(" ")[1],
            thirdy: numbers.split(" ")[2],
            fourty: numbers.split(" ")[3]
        }
    }

    componentDidUpdate(prevProps) {
        if (this.props.rotateCard !== prevProps.rotateCard) {
            this.setState({cardRotete: this.props.rotateCard});
        }

        if (this.props.backImage !== prevProps.backImage) {
            this.setState({changeCard: this.props.backImage});
        }
    }

    render() {
        const {cardNumber, cardName, cvv, shelfLife} = this.props

        const rotete = cx({
            'card__container': true,
            'rotate': this.state.cardRotete
        });

        const changeCardImage = cx({
            'card__front': true,
            'card-imge-complete-front': this.state.changeCard
        });

        const changeCardImageBack = cx({
            'card__back': true,
            'card-imge-complete-back': this.state.changeCard
        });

        return (
            <div className="card">
                <div className={rotete}>
                    <div className={changeCardImage}>
                        <div className="card__front-container">
                            <div className="card__digit-container">
                                <span className="card__digit">
                                    {cardNumber ?
                                    this.formatCardNumber(cardNumber).first : '****'}
                                </span>

                                <span className="card__digit">
                                    {cardNumber && this.formatCardNumber(cardNumber).second ?
                                    this.formatCardNumber(cardNumber).second : '****'}
                                </span>

                                <span className="card__digit">
                                    {cardNumber && this.formatCardNumber(cardNumber).thirdy ?
                                    this.formatCardNumber(cardNumber).thirdy : '****'}
                                </span>

                                <span className="card__digit">
                                    {cardNumber && this.formatCardNumber(cardNumber).fourty ?
                                    this.formatCardNumber(cardNumber).fourty : '****'}
                                </span>

                            </div>
                            <div className="card__titular-info-wrapper">
                                <p className="card__titular-info">{ cardName || 'NOME DO TITULAR' }</p>
                                <p className="card__titular-info">{ shelfLife || '99/99' }</p>
                            </div>
                        </div>
                    </div>
                    <div>
                        <div className={changeCardImageBack}>
                            <div className="card__back-cvv-container">
                                <span className="card__back-cvv">{ cvv || '***' }</span>
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        )
    }
}

Card.propTypes = {
    cardNumber: PropTypes.string,
    cardName: PropTypes.string,
    shelfLife: PropTypes.string,
    cvv: PropTypes.string,
    portionQuantity: PropTypes.number,
    rotateCard: PropTypes.bool,
    backImage: PropTypes.bool,
}

const mapStateToProps = (state) => {
    return {
        cardNumber: state.cardNumber,
        cardName: state.cardName,
        shelfLife: state.shelfLife,
        cvv: state.cvv,
        portionQuantity: state.portionQuantity,
        rotateCard: state.rotateCard,
        backImage: state.backImage,
    }
  }

export default connect(mapStateToProps)(Card);