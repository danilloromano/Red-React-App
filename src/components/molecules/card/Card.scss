@import "../../../assets/variables/variables.scss";

.card {
    width: 280px;
    height: 175px;
    perspective: 1000px;

    &__container {
        position: relative;
        transition: transform 0.6s;
        transform-style: preserve-3d;
        width: 100%;
        height: 100%;
    }

    &__front, &__back {
        position: absolute;
        width: 100%;
        height: 100%;
        -webkit-backface-visibility: hidden;
        backface-visibility: hidden;
        background-repeat: no-repeat;
        background-size: 100% 100%;
    }
    
    &__front-container {
        padding: 80px 14px 0;
    }

    &__front {
        background-image: url('../../../assets/images/card-default.png');
    }
    
    &__back {
        transform: rotateY(180deg);
        background-image: url('../../../assets/images/back-card-default.png');

        &-cvv-container {
            position: relative;
            top: 83px;
            left: 48%;
        }

        &-cvv {
            font-size: 14px;
            color: $black;
            text-shadow: 0px 1px 2px $allBlack;
            letter-spacing: 5px;
        }
    }
    
    &__digit-container {
        display: flex;
        align-items: center;
        justify-content: space-between;
    }

    &__digit {
        font-size: 16px;
        color: #FFFFFF;
        text-shadow: 0px 1px 2px $allBlack;
        letter-spacing: 5px;
    }

    &__titular-info-wrapper {
        margin-top: 25px;
        display: flex;
        align-items: center;
        justify-content: space-between;
    }

    &__titular-info {
        font-size: 16px;
        text-transform: uppercase;
        color: #FFFFFF;
        text-shadow: 0px 1px 2px $allBlack;
    }


}

.rotate {
    transform: rotateY(180deg);
}

.card-imge-complete-front {
    background-image: url('../../../assets/images/card.png');
}

.card-imge-complete-back {
    background-image: url('../../../assets/images/back-card.png');
}

@media screen and (min-width: 1024px) {
    .card {
        width: 365px;
        height: 225px;

        &__front-container {
            padding: 105px 27px 0;
        }

        &__digit {
            font-size: 24px;
        }
    }
}