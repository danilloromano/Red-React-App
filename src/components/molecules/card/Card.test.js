import React from 'react';
import toJson from 'enzyme-to-json'
import { shallow } from 'enzyme';
import Card from './Card';
import configureMockStore from 'redux-mock-store';

describe('Card render test', () => {

    const mockStore = configureMockStore();

    const initialState = {
        cardNumber: "",
        cardName: "",
        shelfLife: "",
        cvv: "",
        portionQuantity: 0,
        rotateCard: false,
        backImage: false,
      }
        const store = mockStore(initialState);

    it('Card should render correctly', () => {
        const wrapper = shallow(<Card store={store}/>)
        expect(toJson(wrapper)).toMatchSnapshot()
    });
});