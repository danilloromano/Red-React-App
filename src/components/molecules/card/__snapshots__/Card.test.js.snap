// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card render test Card should render correctly 1`] = `
<ContextProvider
  value={null}
>
  <Card
    backImage={false}
    cardName=""
    cardNumber=""
    cvv=""
    dispatch={[Function]}
    portionQuantity={0}
    rotateCard={false}
    shelfLife=""
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;

exports[`Card render test Card should render correctly 2`] = `
<ContextProvider
  value={null}
>
  <Card
    backImage={false}
    cardName=""
    cardNumber=""
    cvv=""
    dispatch={[Function]}
    portionQuantity={0}
    rotateCard={false}
    shelfLife=""
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;

exports[`Home render test Home should render correctly 1`] = `
<ContextProvider
  value={null}
>
  <Card
    backImage={false}
    cardName=""
    cardNumber=""
    cvv=""
    dispatch={[Function]}
    portionQuantity={0}
    rotateCard={false}
    shelfLife=""
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />
</ContextProvider>
`;
